### ISSUE TITLE
grub4dos 0.4.6a crashes when it 'sees' a certain partition structure

### ISSUE THREAD

**Issue by steve6375** [chenall/grub4dos#119]
I have had about 6 reports in a year now from different people. Boot from USB grub4dos drive to install linux to server - reboot to USB grub4dos drive - grub4dos goes 'crazy'.
Most of the reports have been when installing linux to a RAID server, so there have been a number of disks in an array, but  the BIOS sees them as separate disks (hd1, hd2, hd2 etc).
When you then boot to grub4dos from a USB drive, when grub4dos tries to access the disks (e.g. find command or similar), it crashes/hangs/loses current disk, etc.
The symptom in E2B is that grub4dos cannot find an E2B.cfg file which is on the USB drive.
See one report https://www.netzgewitter.com/2016/06/easy2boot-e2b-cfg-is-missing/
If /grldr is replaced by a 0.4.5c version, it works fine.
This issue is very hard to replicate (I have not seen it myself), and re-tracing the steps (e.g. re-installing linux on the server) usually results in not seeing the issue again!
I suspect grub4dos is seeing a raw unformatted filesystem (maybe a swap partition?) and crashing but this is a guess!

**Comment by steve6375** [comment]
Thanks to Norbert Harrer, we can now reproduce this problem.
Please download https://www.netzgewitter.com/files/public/E2BTest_VBox.zip
It contains a VirtualBox virtual machine.
If you boot to grub4dos 0.4.6a with the E2BTest.vdi loaded as disk 1 (and USB drive as disk 0), then if you boot to grub4dos and execute from command line...
`geometry (hd1)`
then grub4dos crashes!
There is no problem if you use grub4dos 0.4.5c

Norbert says that if he makes the partition smaller, then it is OK.

> Right. I did some more tests beforehand though. I removed partitions one by one from the partition table. That lead to the fact that only the third partition caused the problem. The third partition is unfortunately the largest partition with about 930GB. So I created a new disk with a single small partition which contained only the troublesome beginning 64KB of the former third partition. Here I found out that the problem goes away when the partition is too small. I don't know the size limit though. The problem was showing up again, once I set the partition to the former size again.
> 
> I was now even able to reproduce the whole thing inside a VirtualBox VM. I created a sparse vdi of about 1TB. Since it's almost empty though, it only uses a couple MB on the host machine. Once E2B is booted with Virtual Machine USB Boot inside VirtualBox, the error message appears.
> 
> Here is a copy of the virtual machine (150KB compressed): https://www.netzgewitter.com/files/public/E2BTest_VBox.zip
> 
> And here are the MBR and the first 64KB of the partition (the drive in the VM already contains a disk with that data though):
> 
> MBR: https://www.netzgewitter.com/files/public/mbr_v4.bin
> First 64K of partition: https://www.netzgewitter.com/files/public/part_start.bin﻿

**Comment by steve6375** [comment]
![image](https://cloud.githubusercontent.com/assets/1469350/16815526/0b4fb3d4-4932-11e6-8a7c-7d50b49ec735.png)

http://www.rmprepusb.com/tutorials/how-to-boot-from-usb-under-vmware-server#TOC-The-solution

**Comment by steve6375** [comment]
![image](https://cloud.githubusercontent.com/assets/1469350/16815686/b67b5182-4932-11e6-9921-dad1361b5541.png)

**Comment by steve6375** [comment]
![image](https://cloud.githubusercontent.com/assets/1469350/16815748/f233639a-4932-11e6-80e2-b9036ea0b13f.png)

The UP key no longer shows previous commands after executing  geometry (hd1) command.
Also, batch files do not run.

**Comment by steve6375** [comment]
![image](https://cloud.githubusercontent.com/assets/1469350/16816065/5f6f86ea-4934-11e6-95e9-05f86d864c0d.png)
The command
if exist (hd1,0)/xxx.yyy echo ddd
also causes problems.
I think memory is being corrupted? Command stack is lost and batch files and .cfg files don't work properly after accessing disk 1.

**Comment by steve6375** [comment]
```
!BAT
# get larger screen
#graphicsmode -1 1024
debug 3
debug msg=3
ls /_ISO/e2b/grub/e2b.cfg
echo
echo 1
echo
geometry (hd1)
echo 2
ls /_ISO/e2b/grub/e2b.cfg
echo 3
```

![image](https://cloud.githubusercontent.com/assets/1469350/16816838/d626d010-4937-11e6-8941-0b66d0ab2df3.png)

If use graphicsmode -1 1024, then no text is visible after geometry command.

![image](https://cloud.githubusercontent.com/assets/1469350/16816896/1492ec94-4938-11e6-9331-84c9119cd244.png)

**Comment by steve6375** [comment]
Are you able to reproduce the problem?

**Comment by yaya2007** [comment]
VirtualBox VMUBDrive000.vmdk failed to mount

**Comment by steve6375** [comment]
You must run VBox as Administrator - use VMUB application to run VBOX or use Run as Administrator

**Comment by yaya2007** [comment]
Open the virtual disk file D:\E2BTest\VMUBDrive000.vmdk failed. 
[VMUBDrive000.vmdk.txt](https://github.com/chenall/grub4dos/files/388347/VMUBDrive000.vmdk.txt)

**Comment by steve6375** [comment]
You must install VMUB application
http://www.rmprepusb.com/tutorials/how-to-boot-from-usb-under-vmware-server#TOC-The-solution
https://github.com/DavidBrenner3/VMUB

Then run VMUB and Add a Vbox VM
Then run VMUB - Start
[image: Inline images 1]
![image](https://cloud.githubusercontent.com/assets/1469350/17212670/82090d80-54c8-11e6-9ee6-d71486232bea.png)

Much faster that QEMU!

**Comment by yaya2007** [comment]
"Latest VMUB v1.6 Download from GitHub" need to compile their own?

**Comment by steve6375** [comment]
https://drive.google.com/open?id=0ByKxAD_t9uvLYUNadFFTdWJaeE0

Link at top of page --> download compiled version

**Comment by yaya2007** [comment]
"x86_VMUB_INNO_script.iss" how to run?

**Comment by steve6375** [comment]
Just download the exe and install it
[image: Inline images 1]

https://drive.google.com/file/d/0ByKxAD_t9uvLYUNadFFTdWJaeE0/view

**Comment by yaya2007** [comment]
"google" cannot access

**Comment by steve6375** [comment]
Can you access MS OneDrive?

**Comment by steve6375** [comment]
https://1drv.ms/u/s!AqlrQcdsFA-KmGeLP8zxpbsMUI0f

**Comment by yaya2007** [comment]
You can add the txt suffix to the file name, and then upload it to the table.

**Comment by yaya2007** [comment]
"https://1drv.ms/u/s!AqlrQcdsFA-KmGeLP8zxpbsMUI0f" cannot access .

**Comment by steve6375** [comment]
https://1drv.ms/t/s!AqlrQcdsFA-KmGgv0HqUqvUOYEp6

**Comment by steve6375** [comment]
I have uploaded it to my website, try
http://files.easy2boot.com/200001818-5bbb25cb48/Virtual%20Machine%20USB%20Boot%201.6%20compiled.zip.txt
Steve

**Comment by yaya2007** [comment]
Downloading

**Comment by yaya2007** [comment]
After download, use the text immediately open, can not be saved.

**Comment by steve6375** [comment]
Same link is on this page
http://www.easy2boot.com/testing-e2b/using-qemu-or-virtual-box/?ez_view_original_only=on
[image: Inline images 1]

**Comment by yaya2007** [comment]
download OK! Thank you

**Comment by yaya2007** [comment]
run:
ls (hd1,0)/
retum: 
Cannot mount selected partition

Is it the same with you?

**Comment by steve6375** [comment]
Yes - that is expected - but try  Up Arrow afterwards - no command stack!

Also try running from a menu.lst or  grub4dos batch file - exits or errors!

Memory is corrupted when you access that partition

**Comment by yaya2007** [comment]
[grldr.rar.txt](https://github.com/chenall/grub4dos/files/391705/grldr.rar.txt)

**Comment by steve6375** [comment]
Seems to work OK.
I will test further to make sure nothing else is broken, but it certainly seems to fix the problem.
Thank you - great work!

**Comment by steve6375** [comment]
Small comment (just cosmetic, not important) - on boot, I get these messages

![image](https://cloud.githubusercontent.com/assets/1469350/17268582/02e23888-5626-11e6-82bf-0cc3e680a2ed.png)

Is there any reason for the line to be shown twice?

**Comment by steve6375** [comment]
**BUG** in new version 2016-07-39  - do not release yet please!
If load XP ISO into memory (map --mem) then it crashes - 0.4.5c is OK
I test with Easy2Boot - XP SP3 ISO - Step 1
Will re-test with old 0.4.6a in few minutes.

**Comment by steve6375** [comment]
Older 0.4.6a 2016-04-07 also fails in same way (I am using a Lenovo IdeaPad 300 to test - EeePC 904HA also fails in same way), but 0.4.5c 2015-05-18 is OK. So this is a different problem. I will report as different issue.
To reproduce, make Easy2Boot USB drive - add XP SP3 ISO to _ISO\XP folder.
Boot to Easy2Boot - CTRL-I - XP Step 1 - use defaults (2, Y=load into memory) - hangs after red text.
More info - 603MB XP ISO is OK, 639MB ISO crashes. Needs more investigation but problem not related to latest bugfix.

**Comment by steve6375** [comment]
Fixed in 2016-08-06 0.4.6a
